libraries/UR_Atmosphere/UR_Atmosphere.cpp

```diff
@@ -0,0 +1,222 @@
+   Altitude_difference calculations was based on the AviationCalculator 1.8.2 05/10/2017
+   revision application from Ph.D. Aerospace Engineering, Joachim K. Hochwarth (GE Principal
+   Engineer/System Architect), published on this URL:
+
+   http://www.hochwarth.com/misc/AviationCalculator.html
+   http://www.hochwarth.com/script/AviationCalculator.js
+
+   Thanks Hochwarth!
+
+   These functions was tested on the Encoder & Altimeter Benchtest Hardware and Application
+   development in cooperation and support with Fenix Aeronautics Academy under the supervision
+   and technical support with the FAA Licensed Fixed and Rotor Wing Aircraft Engineer
+   Maintenance Service, Rudy Averanga <raveranga@fenixaeronautica.com>.
+
+   Thanks a lot Rudy!
+*/
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Baro/AP_Baro.h>
+#include "UR_Atmosphere.h"
+
+extern const AP_HAL::HAL& hal;
+bool UR_Atmosphere::_initialized = false;
+UR_Atmosphere *UR_Atmosphere::_instance = nullptr;
+
+UR_Atmosphere::UR_Atmosphere()
+{
+    _barometer = AP_Baro::get_singleton();
+}
+
+void UR_Atmosphere::init(void)
+{
+    if (_initialized) {
+        return;
+    }
+
+    _initialized = true;
+    _instance = this;
+
+    if (!_barometer) {
+        _barometer = new AP_Baro;
+        _barometer->init();
+    }
+}
+
+void UR_Atmosphere::update(uint8_t sensor)
+{
+    if (!atm_status.updated) {
+        atm_status.updated = true;
+        atm_status.altitude_amsl = calculate_altitude_difference(press0, _barometer->sensors[sensor].pressure, _barometer->sensors[sensor].temperature);
+        atm_status.qnh_pressure = calculate_qnh(atm_status.altitude_amsl, (_barometer->sensors[sensor].pressure / 100), _barometer->sensors[sensor].temperature, ALTITUDE_UNIT::METER);
+        atm_status.qfe_pressure = calculate_qfe(atm_status.altitude_amsl, atm_status.qnh_pressure, _barometer->sensors[sensor].temperature, ALTITUDE_UNIT::METER) * 100;
+    }
+}
+
+// Get the calculation altitude difference based on th Layer 0 > Troposphere.
+float UR_Atmosphere::calculate_altitude_difference(float base_pressure, float pressure, float temp, altitude_unit_t alt_diff_unit) const
+{
+    const float tempcorrected  = Temp0 + temp - 15;
+    float unit_result = 1;
+
+    switch (alt_diff_unit) {
+    case FEET:
+        break;
+    case METER:
+        unit_result = CftTOm;
+        break;
+    default:
+        break;
+    }
+
+    const float result = unit_result * (tempcorrected * (powf((base_pressure / pressure), (dTdh0SI / CgRGasSI)) - 1.0f) / dTdh0);
+
+    return result;
+}
+
+// Calculate QNH by hPa and feet by default.
+// [1] Meteorology Specification A2669 Version 4.00 Page 100
+float UR_Atmosphere::calculate_qnh(float alt_qnh, float pressure_qnh, float temp, altitude_unit_t alt_qnh_unit) const
+{
+    uint8_t pQNHUnit = PRESSURE_UNIT::HECTO_PASCAL;
+    uint8_t resultQNHUnit = PRESSURE_UNIT::HECTO_PASCAL;
+    float hQFE = 0;
+    float pQFE = 0;
+
+    switch (alt_qnh_unit) {
```

Comment:
I can't see what this does for you at all. We only have one use case, this is pointless.

Reply:
Yes, the purpose about it for future extension to others Atm layers like the Stratosphere. Please keep it.